libvast/vast/flatbuffer.hpp

```diff
@@ -0,0 +1,350 @@
+  flatbuffer() noexcept = default;
+
+  /// Constructs a ref-counted FlatBuffers root table from a FlatBufferBuilder
+  /// by finishing it.
+  // TODO: FlatBuffers 2.0's reflection allows for inferring the
+  // *file_identifier* from *Table*, which makes this API a lot easier to use.
+  // When upgrading to that version from 1.12 we can enable reflection when
+  // compiling the schemas.
+  flatbuffer(flatbuffers::FlatBufferBuilder& builder,
+             flatbuffers::Offset<Table> offset,
+             const char* file_identifier) noexcept requires(IsRootTable) {
+    builder.Finish(offset, file_identifier);
+    auto chunk = chunk::make(builder.Release());
+    VAST_ASSERT(chunk);
+    *this = flatbuffer{chunk};
+  }
+
+  /// Constructs a ref-counted FlatBuffers table that shares the
+  /// lifetime with another FlatBuffer pointer.
+  /// @pre `parent`
+  /// @pre *table* must be accessible from *parent*.
+  template <class ParentTable, bool ParentIsRootTable>
+  flatbuffer(flatbuffer<ParentTable, ParentIsRootTable> parent,
+             const Table& table) noexcept requires(!IsRootTable)
+    : chunk_{std::exchange(parent.chunk_, {})}, table_{&table} {
+    VAST_ASSERT(chunk_);
+    VAST_ASSERT(reinterpret_cast<const std::byte*>(table_) >= chunk_->data());
+    VAST_ASSERT(reinterpret_cast<const std::byte*>(table_)
+                < (chunk_->data() + chunk_->size()));
+  }
+
+  ~flatbuffer() noexcept = default;
+
+  flatbuffer(const flatbuffer& other) noexcept
+    : chunk_{other.chunk_}, table_{other.table_} {
+    // nop
+  }
+
+  flatbuffer(flatbuffer&& other) noexcept
+    : chunk_{std::exchange(other.chunk_, {})},
+      table_{std::exchange(other.table_, {})} {
+    // nop
+  }
+
+  flatbuffer(std::nullptr_t) noexcept {
+    // nop
+  }
+
+  flatbuffer& operator=(const flatbuffer& rhs) noexcept {
+    if (&rhs == this)
+      return *this;
+    chunk_ = rhs.chunk_;
+    table_ = rhs.table_;
+    return *this;
+  }
+
+  flatbuffer& operator=(flatbuffer&& rhs) noexcept {
+    chunk_ = std::exchange(rhs.chunk_, {});
+    table_ = std::exchange(rhs.table_, {});
+    return *this;
+  }
+
+  flatbuffer& operator=(std::nullptr_t) noexcept {
+    chunk_ = nullptr;
+    table_ = nullptr;
+    return *this;
+  }
+
+  // -- operators -------------------------------------------------------------
+
+  explicit operator bool() const noexcept {
+    return table_ != nullptr;
+  }
+
+  const Table& operator*() const noexcept {
+    VAST_ASSERT(table_);
+    return *table_;
+  }
+
+  const Table* operator->() const noexcept {
+    VAST_ASSERT(table_);
+    return table_;
+  }
+
+  friend bool operator==(flatbuffer lhs, flatbuffer rhs) noexcept
+    requires(IsRootTable) {
+    if (&lhs == &rhs)
+      return true;
+    if (lhs && rhs) {
+      const auto lhs_bytes = as_bytes(lhs);
+      const auto rhs_bytes = as_bytes(rhs);
+      return std::equal(lhs_bytes.begin(), lhs_bytes.end(), rhs_bytes.begin(),
+                        rhs_bytes.end());
```

Comment:
I'm not sure that a user would expect byte equality here, the closest equivalent is `shared_ptr` which only does pointer equality.

Reply:
I intentionally named this `flatbuffer<T>` rather than `flatbuffer_ptr<T>` since this supposed to be use like a shared buffer, not a shared pointer like type.